docs/sphinx/hpx-functions.rst

```diff
@@ -0,0 +1,102 @@
+.. _HPX Functions:
+
+=============
+HPX Functions
+=============
+
+This document describes the prototypes, the allowable arguments of the listed
+HPX functions.
+
+async
+=====
+
+Async in C++ allows running asynchronous programming. Async programming is useful
+in compute heavy implementations. Basic async function would look like
+
+.. code-block:: text
+   #include <future>
+   using namespace hpx;
+   int main() {
+    future<int> f = async([]() {});
+    int val = f.get();
+   }
+
+You can chain multiple async operations. For implementing chain async programming
+``objFuture.then(...).then(...).then(...).then(...).then(...)``.
+
+Implementing async functions should have input parameters as
+
+- Lambda function
+- Executor function
+- Launch policy
+
+then
+====
+
+then in asychronous programming is used in chained async implementations. Using
+then function will avoid blocking waits or wasting threads on polling. Future
+can have then chained implementations. With then the antecedent future is ready
+(has a value or exception stored) before the continuation starts as instructed
+by lambda function. Example implementation using hpx is
+
+.. code-block:: text
+
+   #include <future>
+
+   int main() {
+     future<int> fA = async([])(){});
```

Comment:
Same as above, the return type of the function must match the type held by the returned future.

Reply:
8ad37ee.